File: Arrays/subArraySort.js
const subArraySort = (array) => {
    let minUnsorted = Infinity
    let maxUnsorted = -Infinity
    for (let i = 0; i < array.length - 1; i++) {
        if (i === 0 && array[i] > array[i + 1] ||
            i === array.length - 1 && array[i] < array[i - 1] ||
            array[i] > array[i + 1]) {
                minUnsorted = Math.min(minUnsorted, array[i + 1])
                maxUnsorted = Math.max(maxUnsorted, array[i])
            }
    }
    if (minUnsorted === Infinity) {
        return [-1, -1]
    }
    let left = 0
    let right = array.length - 1

    while (array[left] <= minUnsorted) {
        left++
    }
    while (array[right] >= maxUnsorted) {
        right --
    }
    return [left, right]
}

File: Arrays/subarray_sort.py
def subarray_sort(array):
    minUnsorted = float("inf")
    maxUnsorted = float("-inf")

    for i in range(len(array)):
        num = array[i]
        if out_of_order(i, num, array):
            minUnsorted = min(minUnsorted, num)
            maxUnsorted = max(maxUnsorted, num)
    
    if minUnsorted == float("inf"):
        return [-1, -1]
    
    left = 0
    right = len(array) - 1

    while array[left] <= minUnsorted:
        left += 1
    while array[right] >= maxUnsorted:
        right -= 1
    
    return [left, right]

def out_of_order(i, num, array):
    if i == 0:
        return num > array[i + 1]
    if i == len(array) - 1:
        return num < array[i - 1]
    return num > array[i + 1] or num < array[i - 1]